src/atomic/atomic_cell.rs

```diff
@@ -0,0 +1,695 @@
+    where
+        F: FnMut(T) -> T,
+    {
+        let mut current = self.get();
+
+        loop {
+            let new = f(current);
+
+            let previous = unsafe {
+                atomic_compare_and_swap(self.value.get(), current, new)
+            };
+
+            if byte_eq(&previous, &current) {
+                return new;
+            }
+
+            current = previous;
+        }
+    }
+}
+
+impl<T: Copy + Eq> AtomicCell<T> {
+    /// If the current value equals `current`, stores `new` into the atomic cell.
+    ///
+    /// Returns `true` if the value was updated, and `false` otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.compare_and_set(1, 8), false);
+    /// assert_eq!(a.get(), 7);
+    ///
+    /// assert_eq!(a.compare_and_set(7, 8), true);
+    /// assert_eq!(a.get(), 8);
+    /// ```
+    pub fn compare_and_set(&self, mut current: T, new: T) -> bool {
+        loop {
+            let previous = unsafe {
+                atomic_compare_and_swap(self.value.get(), current, new)
+            };
+
+            if byte_eq(&previous, &current) {
+                return true;
+            }
+
+            if previous != current {
+                return false;
+            }
+
+            // Since `byte_eq(&previous, &current)` is `false`, that means the compare-and-swap
+            // operation failed and didn't store `new`. However, `previous == current`, which means
+            // it technically should've succeeded.
+            //
+            // We cannot return neither `true` nor `false` here because the operation didn't
+            // succeed nor fail, but simply encountered an inconsistent state. The only option left
+            // is to retry with `previous` as the new `current`.
+            current = previous;
+        }
+    }
+}
+
+macro_rules! impl_arithmetic {
+    ($t:ty, $example:tt) => {
+        impl AtomicCell<$t> {
+            /// Increments the inner value by `val` and returns the new value.
+            ///
+            /// The addition wraps on overflow. Note that `atomic_cell.add(val)` is equivalent to:
+            ///
+            /// ```ignore
+            /// atomic_cell.update(|x| x.wrapping_add(val))
+            /// ```
+            ///
+            /// # Examples
+            ///
+            /// ```
+            /// use crossbeam_utils::atomic::AtomicCell;
+            ///
+            #[doc = $example]
+            ///
+            /// assert_eq!(a.add(3), 10);
+            /// assert_eq!(a.get(), 10);
+            /// ```
+            #[inline]
+            pub fn add(&self, val: $t) -> $t {
+                if can_transmute::<$t, atomic::AtomicUsize>() {
+                    let a = unsafe { &*(self.value.get() as *const atomic::AtomicUsize) };
+                    a.fetch_add(val as usize, Ordering::SeqCst).wrapping_add(val as usize) as $t
```

Comment:
I'm not a fan of using `SeqCst` in loads and stores.  I described the reason here: https://github.com/rust-lang/rust/pull/31650#issuecomment-327333760  In short, as contrary to commonly believed, using `SeqCst` for loads, stores, and rmws doesn't give you a useful guarantee over `Release`/`Acquire` semantics.  In particular, it has never been proved that using `SeqCst` for a location gives you a perfect ordering for that particular location.

For `AtomicCell`, it seems `Release` and `Acquire` will be enough, I think.

Reply:
I feel I still haven't fully grasped the consequences of the broken C/C++11 model. :( Maybe we should take a look at an example. Here's one with 4 threads:

```rust
// THREAD 1
a.set(1);

// THREAD 2
b.set(1);

// THREAD 3
a.get(); // 1
b.get(); // 0

// THREAD 4
b.get(); // 1
a.get(); // 0
```

Now, two questions:

1. Is this scenario possible if all `get` operations use `Acquire` and all `set` operations use `SeqCst`? I would say "yes".

2. Is this scenario possible if all `get` and all `set` operations use `SeqCst`? I would say "no" because the model prohibits this behavior. But you're saying the answer is "no" because the compilers have no way of compiling this correctly on Power/ARM? Is my understanding correct?